### ISSUE TITLE
Connect options no longer accept user & pass as empty strings

### ISSUE THREAD

**Issue by mleanos** [Automattic/mongoose#3484]
Previously, in version `4.1.11` I was able to pass the `user` and `pass` options into the `connect` method as empty strings. Like so..
```javascript
var db = mongoose.connect('mongodb://localhost/dev', { user: '', pass: '' }, function (err) {
  if (err) {
    console.log(err);
  } else {
    console.log('connected');
  }
});
```
However, in version `4.1.12`, this no longer works. I get a `auth failed` error. I have verified that there's a difference in the way the authentication is handling the `user` and `pass` between the two versions of this library.

```javascript
// from 4.1.11
// test passes
it('including user/pass as empty strings', function(done) {
      var db = mongoose.createConnection('localhost', 'fake', 27000, {user: '', pass: ''});
      db.on('error', function() {});
      assert.equal('object', typeof db.options);
      assert.equal('object', typeof db.options.server);
      assert.equal(true, db.options.server.auto_reconnect);
      assert.equal('object', typeof db.options.db);
      assert.equal(false, db.options.db.forceServerObjectId);
      assert.equal('fake', db.name);
      assert.equal('localhost', db.host);
      assert.equal(27000, db.port);
      assert.equal(undefined, db.pass);
      assert.equal(undefined, db.user);
      db.close();
      done();
    });

// from 4.1.12
// test fails on the db.shouldAuthenticate() assertion. 
describe('when username and password are empty strings', function() {
        it('should return false', function(done) {
          var db = mongoose.createConnection('localhost', 'fake', 27000, { user: '', pass: ''});
          db.on('error', function() {});
          assert.equal('object', typeof db.options);
          assert.equal('object', typeof db.options.server);
          assert.equal(true, db.options.server.auto_reconnect);
          assert.equal('object', typeof db.options.db);
          assert.equal(false, db.options.db.forceServerObjectId);
          assert.equal('fake', db.name);
          assert.equal('localhost', db.host);
          assert.equal(27000, db.port);
          assert.equal('', db.pass);
          assert.equal('', db.user);

          assert.equal(false, db.shouldAuthenticate()); // db.shouldAuthenticate() returns true

          db.close();
          done();
        });
      });
```

If you notice that in `4.1.11` the `user` and `pass` options are passed back as undefined. This tells me that they are being deliberately handled & bypassed by the authentication. `4.1.12` doesn't handle them the same. They are passed back as their original empty strings.

Is the behavior in `4.1.12` a bug, or does it fix a previous bug? Or perhaps this is just a purposed change in the way authentication is treating use cases as this.

**Comment by vkarpov15** [comment]
Might have something to do with the x509 bugfix, #3454. Will take a look at this in the near future.

**Comment by mleanos** [comment]
@vkarpov15 The problem seems to be here... https://github.com/Automattic/mongoose/blob/4.1.12/lib/connection.js#L731-L732

Since this is checking against `null`, when empty strings are provided for either the `user` and `pass` options then it will assume the options due indeed provide the authentication data; thus, passing empty strings along to the internal driver connection, resulting in a failed authentication.

The question remains.. Is this a bug, or the desired behavior? 

There are cases when an empty string being passed for the user/pass may be appropriate, and still allow a connection to the database. For instance, in a local environment where the database doesn't require authentication.

This patch/fix/workaround seems to revert the behavior back to it's previous state. Although, I don't think the checks here should be checking for `null`. 
https://github.com/mleanos/mongoose/commit/90218c0fa92224a252bcbb0532b9f653b6f397d0

Do you think this is a viable solution, or would you rather a user of this library *remove* the user/pass options before attempting to connect? I wouldn't see a problem with having Mongoose be responsible for reconciling empty strings. It seems this would fall in line with the purpose of the `optionsProvideAuthenticationData` method.

Thoughts?

**Comment by vkarpov15** [comment]
Closed re: #3498. Thanks a lot for your contribution!

**Comment by mleanos** [comment]
@vkarpov15 You're welcome! It was my pleasure.

When can we expect this to be released? Since `4.2.1` was already released.

**Comment by vkarpov15** [comment]
4.2.3 is scheduled to be released tomorrow.
